### ISSUE TITLE
Questions about loss

### ISSUE THREAD

**Issue by K-Mike** [UKPLab/sentence-transformers#622]
I just watched your repo a little bit and don't understand some moments:
1. In  CosineSimilarityLoss you use MSE(CosineSimilarity (v0, v1)), why so? 
2. Why don't your use CosineEmbeddingLoss?
3. Do you use  a linear layer inside loss class, why do you choose this way?
4. You use (u, v, |u - v|) but check Cosine Similarity everywhere, why?
5. I think your compare different loss functions, have you public the results?
Thank you for your work!

**Comment by nreimers** [comment]
@K-Mike 
1+2) CosineEmbeddingLoss is only defined for y=1 and y=-1. But the gold score can also have values like 0.2 or 0.8 or 0.5
3+4) The models are trained on NLI data, which is a 3-way classification. Hence, we need 3-way softmax classifier in the the softmax loss. As input for the classifier, we use  (u, v, |u - v|) .
5) The loss function largely depends on the type of data / labels you have. Not every loss function is suitable for every data you have. So choosing the loss mainly depends what types of labels you have and what your task is. Hence, comparing loss functions is only possible in a limited way.

**Comment by K-Mike** [comment]
Thank you!

**Comment by liebkne** [comment]
@K-Mike May I ask one question? If the labels range from [0, 1], does one need to project the model output to the same range before [applying MSE](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/losses/CosineSimilarityLoss.py#L41) ?

**Comment by nreimers** [comment]
You don't have to do this.

**Comment by liebkne** [comment]
@nreimers Thank you!

**Comment by giovcandido** [comment]
> @K-Mike 1+2) CosineEmbeddingLoss is only defined for y=1 and y=-1. But the gold score can also have values like 0.2 or 0.8 or 0.5 3+4) The models are trained on NLI data, which is a 3-way classification. Hence, we need 3-way softmax classifier in the the softmax loss. As input for the classifier, we use (u, v, |u - v|) . 5) The loss function largely depends on the type of data / labels you have. Not every loss function is suitable for every data you have. So choosing the loss mainly depends what types of labels you have and what your task is. Hence, comparing loss functions is only possible in a limited way.

Isn't CosineEmbeddingLoss defined for y = 1 (positive pair) and y = -1 (negative pair)? I think y can represent the label. So, this loss could be implemented and work similarly to contrastive loss. What do you think?
